stan/math/prim/meta/is_eigen.hpp

```diff
@@ -33,6 +34,35 @@ struct is_eigen_base<Eigen::EigenBase<T>> : std::true_type {};
 
 }  // namespace internal
 
+namespace internal {
+/*
+ * Underlying implimenation to check if a type is derived from EigenBase
+ */
+template <typename T>
+struct is_eigen_dense_base
+    : std::integral_constant<bool,
+                             std::is_base_of<Eigen::DenseBase<T>, T>::value> {};
+
+template <typename T>
+struct is_eigen_dense_base<Eigen::DenseBase<T>> : std::true_type {};
+
+}  // namespace internal
```

Comment:
This can all go in the same namespace.  

But then our convention has been one function per file with the file named after the function.  Let's ask @syclik what he thinks about changing that pattern.  I'm OK either way.